### PR TITLE
enable zero copy Conversion for OnnxLightGlue via DLPack

### LINKED ISSUE #3696: [Feature]: Reduce redundant tensor layout transformations in Image API
### 🚀 Feature Description

Optimize the kornia.image.Image API by minimizing redundant tensor layout transformations (permute) and enabling more efficient image processing pipelines through canonical layout handling and optional in-place operations.

### 📂 Feature Category

Image Processing

### 💡 Motivation

Currently, multiple methods in the Image class (e.g., to_gray, to_rgb, to_bgr) internally perform repeated layout conversions when handling CHANNELS_LAST inputs.

Typical pattern:

Convert to CHANNELS_FIRST via permute
Apply operation
Convert back via permute

This results in:

Repeated tensor stride changes and potential memory reordering
Increased overhead in chained pipelines
Avoidable data movement, especially for large images or batched inputs

While Kornia already supports zero-copy interop via DLPack, these internal transformations introduce a separate performance bottleneck at the execution level.

### 💭 Proposed Solution

Option 1: Canonical Internal Layout

Standardize internal representation to CHANNELS_FIRST:

Convert once at construction or entry point
Perform all operations in canonical layout
Convert back only when explicitly requested
Option 2: Lazy Layout Handling

Track layout via metadata without immediate permute:

Defer physical layout transformation until required
Avoid unnecessary conversions in chained operations
Option 3: In-place Variants

Introduce in-place APIs to reduce allocations:

img.to_gray_()
img.to_rgb_()
Option 4: Operation Fusion (optional future work)

Enable direct conversions (e.g., BGR → GRAY) without intermediate representations.

### 🔄 Alternatives Considered

Keeping current explicit layout handling (simpler but less efficient)
Relying on users to manually normalize layout before using API (error-prone and not ergonomic)

### 🎯 Use Cases

High-throughput image pipelines (e.g., preprocessing for deep learning)
Robotics / real-time vision systems
Batched image transformations on GPU
Scenarios where minimizing memory movement is critical

### 📝 Additional Context

This issue is conceptually similar to how DLPack eliminates unnecessary memory copies across frameworks. Here, the goal is to reduce intra-framework data movement by optimizing layout handling and execution flow.

Initial profiling suggests that redundant permute operations can contribute significantly to runtime in chained transformations.

### 🤝 Contribution Intent

- [x] I plan to submit a PR to implement this feature
- [ ] I'm requesting this feature but not planning to implement it

### PR DESCRIPTION
**Pull request by Varun-sai-500**
## 📝 Description

Fixes #3696 
Superseeds: #3598 

The original DLPack optimization introduced a guard hasattr(matches, "to_dlpack") that was always False in practice — making the zero-copy path dead code and leaving every CUDA user on the slow NumPy route silently.
Root cause: io_binding.get_outputs() returns ORT's Python-level wrapper (onnxruntime_inference_collection.OrtValue), which does not expose to_dlpack. The DLPack protocol is only available on the underlying pybind11 C object, accessible via ._ortvalue.
This PR fixes the guard to reach through to the C object and enables the zero-copy path that was always intended.

---

## 🛠️ Changes Made
 [x] Replaced hasattr(matches, "to_dlpack") with getattr(matches, "_ortvalue", matches) + hasattr(_m, "__dlpack__") — targets the correct pybind11 C object
 [x] Switched from torch.utils.dlpack.from_dlpack(matches.to_dlpack()) to torch.from_dlpack(_m) — uses the modern PEP 474 __dlpack__ protocol
 [x] Guard is forward-compatible: if a future ORT version exposes __dlpack__ directly on the Python wrapper, getattr falls back to matches itself and the if-block still fires correctly
 [x] Removed unused from torch.utils import dlpack import

---

## 🧪 How Was This Tested?
- [x] **Unit Tests:** (List new/updated tests): existing test_forward covers device correctness on outputs for both CUDA (if-block) and CPU (else-block). Added dtype assertions outputs["matches"].dtype == torch.int64 and outputs["scores"].dtype == torch.float32
- [x] **Manual Verification:** confirmed _ortvalue.__dlpack__ exists and returns a valid CUDA tensor via torch.from_dlpack() on ORT 1.22.0 + Torch 2.11.0 + T4
- [x] **Performance/Edge Cases:**  benchmarked at real LightGlue output shapes (S, 2) int64 + (S,) float32 across S=512→4096. if-block latency is flat ~30µs (pointer handshake, zero PCIe), else-block scales with data size. 2.9x speedup at S=2048 (84µs → 29µs) on Tesla T4

---

## 🕵️ AI Usage Disclosure

- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.

---

## 🚦 Checklist
- [ ] I am assigned to the linked issue (required before PR submission)
- [ ] The linked issue has been approved by a maintainer
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] (Optional) I have attached screenshots/recordings for UI changes.

---

### CODE REVIEW

**Comment by Varun-sai-500** [comment]
image


"""
Benchmark: if-block (DLPack) vs else-block (NumPy) in OnnxLightGlue.forward()
Measures exactly the two code paths as they appear in the fixed PR.

Run:
    pip install onnxruntime-gpu onnx torch -q
    python benchmark_lightglue_conversion.py
"""

import time
import numpy as np
import torch
import onnxruntime as ort

# ── Config ──────────────────────────────────────────────────────────────────
WARMUP  = 100
ITERS   = 1000
DEVICE  = "cuda"

S_VALUES = [512, 1024, 2048, 4096]
# ────────────────────────────────────────────────────────────────────────────


def make_model(S: int) -> bytes:
    from onnx import helper, TensorProto
    import onnx

    ins  = [helper.make_tensor_value_info("matches_in", TensorProto.INT64, [S, 2]),
            helper.make_tensor_value_info("scores_in",  TensorProto.FLOAT, [S])]
    outs = [helper.make_tensor_value_info("matches",    TensorProto.INT64, [S, 2]),
            helper.make_tensor_value_info("scores",     TensorProto.FLOAT, [S])]
    nodes = [helper.make_node("Identity", ["matches_in"], ["matches"]),
             helper.make_node("Identity", ["scores_in"],  ["scores"])]
    graph = helper.make_graph(nodes, "pt", ins, outs)
    model = helper.make_model(graph, opset_imports=[helper.make_opsetid("", 13)])
    model.ir_version = 10
    onnx.checker.check_model(model)
    return model.SerializeToString()


def get_outputs(S: int):
    """Return real IOBinding get_outputs() for a given S."""
    providers = [("CUDAExecutionProvider", {"device_id": 0}), "CPUExecutionProvider"]
    sess = ort.InferenceSession(make_model(S), providers=providers)

    np_m = np.random.randint(0, S, (S, 2), dtype=np.int64)
    np_s = np.random.rand(S).astype(np.float32)

    ov_m = ort.OrtValue.ortvalue_from_numpy(np_m, DEVICE, 0)
    ov_s = ort.OrtValue.ortvalue_from_numpy(np_s, DEVICE, 0)

    binding = sess.io_binding()
    binding.bind_ortvalue_input("matches_in", ov_m)
    binding.bind_ortvalue_input("scores_in",  ov_s)
    binding.bind_output("matches", device_type=DEVICE, device_id=0)
    binding.bind_output("scores",  device_type=DEVICE, device_id=0)
    sess.run_with_iobinding(binding)

    return binding.get_outputs()


def sync():
    torch.cuda.synchronize()


def bench(fn) -> float:
    """Returns mean latency in µs."""
    for _ in range(WARMUP):
        fn()
    sync()
    t0 = time.perf_counter()
    for _ in range(ITERS):
        fn()
    sync()
    return (time.perf_counter() - t0) / ITERS * 1e6


def run():
    try:
        import onnx
    except ImportError:
        raise SystemExit("pip install onnx")
    if not torch.cuda.is_available():
        raise SystemExit("CUDA required.")

    # ── Header ───────────────────────────────────────────────────────────────
    print()
    print("=" * 65)
    print("  OnnxLightGlue output conversion — if-block vs else-block")
    print(f"  Device : {torch.cuda.get_device_name(0)}")
    print(f"  ORT    : {ort.__version__}   Torch : {torch.__version__}")
    print(f"  Warmup : {WARMUP}   Iters : {ITERS}")
    print("=" * 65)
    print()
    print("  Code paths under test:")
    print()
    print("  IF-block   (DLPack, zero-copy):")
    print("    _m = getattr(matches, '_ortvalue', matches)")
    print("    torch.from_dlpack(_m)")
    print()
    print("  ELSE-block (NumPy, double PCIe hop):")
    print("    torch.from_numpy(matches.numpy()).to(self.device)")
    print()

    # ── Sanity check ─────────────────────────────────────────────────────────
    m_chk, _ = get_outputs(64)
    _m = getattr(m_chk, "_ortvalue", m_chk)
    assert hasattr(_m, "__dlpack__"), "_ortvalue has no __dlpack__ — DLPack path unavailable"
    t = torch.from_dlpack(_m)
    assert t.device.type == "cuda"
    print("  [OK] DLPack path confirmed: _ortvalue.__dlpack__ works, tensor on CUDA")
    print()

    # ── Table ─────────────────────────────────────────────────────────────────
    print(f"  {'S (matches)':<16} {'bytes moved':>12} {'else µs':>10} {'if µs':>10} {'speedup':>10}")
    print(f"  {'-' * 62}")

    for S in S_VALUES:
        matches, mscores = get_outputs(S)

        # bytes the NumPy path physically moves across PCIe (x2 for D2H + H2D)
        tensor_bytes = (S * 2 * 8) + (S * 4)   # int64 (S,2) + float32 (S,)
        pcie_bytes   = tensor_bytes * 2          # D2H then H2D

        # ── ELSE block (exact copy from the file) ──
        def else_block():
            torch.from_numpy(matches.numpy()).to(DEVICE)
            torch.from_numpy(mscores.numpy()).to(DEVICE)

        # ── IF block (exact copy from the file) ──
        def if_block():
            _m = getattr(matches, "_ortvalue", matches)
            _s = getattr(mscores, "_ortvalue", mscores)
            torch.from_dlpack(_m)
            torch.from_dlpack(_s)

        t_else = bench(else_block)
        t_if   = bench(if_block)
        speedup = t_else / t_if

        print(f"  {S:<16} {pcie_bytes/1024:>10.1f} KB {t_else:>10.2f} {t_if:>10.2f} {speedup:>9.1f}x")

    print()
    print("  bytes moved = what the else-block physically transfers (D2H + H2D)")
    print("  if-block    = zero bytes moved, pointer handshake only")
    print("=" * 65)
    print()


if __name__ == "__main__":
    run()

File: kornia/feature/lightglue_onnx/lightglue.py [hunk 1]
```diff
@@ -187,10 +186,12 @@ def forward(self, data: dict[str, dict[str, torch.Tensor]]) -> dict[str, torch.T
...
-        if hasattr(matches, "to_dlpack"):
+        _m = getattr(matches, "_ortvalue", matches)
+        _s = getattr(mscores, "_ortvalue", mscores)
+        if hasattr(_m, "__dlpack__"):
             outputs = {
-                "matches": dlpack.from_dlpack(matches.to_dlpack()),
-                "scores": dlpack.from_dlpack(mscores.to_dlpack()),
+                "matches": torch.from_dlpack(_m),
+                "scores": torch.from_dlpack(_s),
             }
```
[the remark this answers is not in the document]
Comment (Varun-sai-500) [reply kornia/feature/lightglue_onnx/lightglue.py]: I added the check

[the remark this answers is not in the document]
Comment (Varun-sai-500) [reply kornia/feature/lightglue_onnx/lightglue.py:207]: I added tests

**Review by edgarriba — APPROVED** [review]
